### PR TITLE
varlink images

### PR DESCRIPTION
**Pull request by baude**
implement varlink image functions for working with libpod with the exception of a
couple due to incompletions on the libpod side of things (build).

also, created a first pass at a libpodpy package which will stand as a client to
working with libpod's varlink methods using python.

Signed-off-by: baude <bbaude@redhat.com>

### CODE REVIEW

**Comment by baude** [comment]
Reviewers, please note a couple of things when reviewing this:
* no new unit or integration tests for varlink because @jwhonce has agreed to do that for me
* the libpodpy skeleton I added will be picked up by @jwhonce as part of the testing. he will likely alter names, etc as he does that inheritance.

File: cmd/podman/commit.go [hunk 1]
```diff
@@ -7,6 +7,7 @@ import (
 	"strings"
 
 	"github.com/pkg/errors"
+	rt "github.com/projectatomic/libpod/cmd/podman/util"
```
Comment (rhatdan) [inline cmd/podman/commit.go]: Why did you not use util?

File: cmd/podman/import.go [hunk 2]
```diff
@@ -7,10 +7,11 @@ import (
 	"net/http"
 	"net/url"
 	"os"
-	"strings"
 
 	"github.com/opencontainers/image-spec/specs-go/v1"
 	"github.com/pkg/errors"
+	cliutil "github.com/projectatomic/libpod/cmd/podman/util"
+	rt "github.com/projectatomic/libpod/cmd/podman/util"
```
Comment (rhatdan) [inline cmd/podman/import.go]: Huh why are you including the same thing twice with two different names?

**Comment by rhatdan** [comment]
Can you rename this directory  cmd/podman/ioprojectatomicpodman to varlink

File: cmd/podman/import.go (hunk 2 above)
Comment (mheon) [reply cmd/podman/import.go]: Better question: Why does Go even allow that without compile errors

File: cmd/podman/ioprojectatomicpodman/io.projectatomic.podman.varlink [hunk 3]
```diff
@@ -17,6 +17,35 @@ type StringResponse (
     message: string
 )
 
+type ImageInList (
```
Comment (mheon) [inline cmd/podman/ioprojectatomicpodman/io.projectatomic.podman.varlink]: Can we get comments on all of these structs in the .varlink? I understand you're working to document the generated Go, but we should also try and provide some docs for the API definition as well.

In particular - what is ImageInList? That name is really confusing

**Comment by rhatdan** [comment]
You are creating a util.go next to a utils.go?  Why?

File: libpod/image/image.go [hunk 4]
```diff
@@ -772,3 +781,18 @@ func splitString(input string) string {
 func (i *Image) InputIsID() bool {
 	return strings.HasPrefix(i.ID(), i.InputName)
 }
+
+// Containers returns the storage containers associated with the image
+func (i *Image) Containers() ([]storage.Container, error) {
```
Comment (mheon) [inline libpod/image/image.go]: Can you change this to return []string with just the container IDs instead? I don't think Varlink needs anything more than a number of containers.

File: libpodpy/__init__.py [hunk 5]
```diff
@@ -0,0 +1,4 @@
+
```
Comment (mheon) [inline libpodpy/__init__.py]: Can you stick all the libpodpy stuff in `contrib/` to distinguish it from the main repo?

File: libpodpy/containers.py [hunk 6]
```diff
@@ -0,0 +1,8 @@
+
+class Containers(object):
+
+    def __init__(self, client):
+        self.client = client
+
+    def List(self):
+        pass
```
Comment (mheon) [inline libpodpy/containers.py]: Missing newline

File: pkg/varlinkapi/config.go [hunk 7]
```diff
@@ -1,14 +1,23 @@
...
 }
 
-var (
-	lp = LibpodAPI{}
-	// VarlinkLibpod instantiation
-	VarlinkLibpod = ioprojectatomicpodman.VarlinkNew(&lp)
-)
+//var (
+//	lp = LibpodAPI{}
+//	// VarlinkLibpod instantiation
```
Comment (mheon) [inline pkg/varlinkapi/config.go]: Can you remove the commented-out code? It looks like it's been replaced by New

File: pkg/varlinkapi/images.go [hunk 8]
```diff
@@ -1,75 +1,285 @@
...
+	}
+	return call.ReplyInspectImage(string(b[:]))
 }
 
-// HistoryImage ...
-func (i *LibpodAPI) HistoryImage(call ioprojectatomicpodman.VarlinkCall) error {
-	return call.ReplyMethodNotImplemented("HistoryImage")
+// HistoryImage returns the history of the image's layers
+// Requires an image or name
+func (i *LibpodAPI) HistoryImage(call ioprojectatomicpodman.VarlinkCall, name string) error {
```
Comment (mheon) [inline pkg/varlinkapi/images.go]: Wait, this returns just `error`... Are the normal `call.ReplyHistoryImage()` responses considered errors? Ew.

File: pkg/varlinkapi/images.go [hunk 9]
```diff
@@ -1,75 +1,285 @@
...
+		return call.ReplyImageNotFound(name)
+	}
+	history, layerInfos, err := newImage.History(getContext())
+	if err != nil {
+		return call.ReplyErrorOccurred(err.Error())
+	}
+	var histories []ioprojectatomicpodman.ImageHistory
+	for i, h := range history {
+		imageHistory := ioprojectatomicpodman.ImageHistory{
+			//Id:
```
[the remark this answers is not in the document]
Comment (baude) [reply pkg/varlinkapi/images.go]: I couldn't find it ... will follow up with you.

File: pkg/varlinkapi/images.go (hunk 8 above)
Comment (baude) [reply pkg/varlinkapi/images.go]: the return is defined in the generated file. it's a bit of a thing to get used to, but this is correct as is.

File: cmd/podman/commit.go (hunk 1 above)
Comment (baude) [reply cmd/podman/commit.go]: Some podman commands have util already imported so I was trying to differentiate.

Comment (baude) [reply cmd/podman/commit.go]: and util is imported later

File: cmd/podman/import.go (hunk 2 above)
Comment (baude) [reply cmd/podman/import.go]: fixed, just an oversight

File: cmd/podman/ioprojectatomicpodman/io.projectatomic.podman.varlink (hunk 3 above)
Comment (baude) [reply cmd/podman/ioprojectatomicpodman/io.projectatomic.podman.varlink]: I will add some comments to describe the structures, sure.

File: libpodpy/containers.py (hunk 6 above)
Comment (baude) [reply libpodpy/containers.py]: done

File: pkg/varlinkapi/config.go (hunk 7 above)
Comment (baude) [reply pkg/varlinkapi/config.go]: done

File: pkg/varlinkapi/images.go [hunk 10]
```diff
@@ -1,75 +1,285 @@
...
+	newImage, err := runtime.ImageRuntime().NewFromLocal(name)
+	if err != nil {
+		return call.ReplyImageNotFound(name)
+	}
+	inspectInfo, err := newImage.Inspect(getContext())
+	b, err := json.Marshal(inspectInfo)
+	if err != nil {
+		return call.ReplyErrorOccurred(fmt.Sprintf("unable to serialize"))
+	}
+	return call.ReplyInspectImage(string(b[:]))
```
[the remark this answers is not in the document]
Comment (baude) [reply pkg/varlinkapi/images.go]: ok

File: libpod/image/image.go (hunk 4 above)
Comment (mheon) [reply libpod/image/image.go]: I'd still prefer not to return storage.Container - the ID should be plenty if you want to work with the container in libpod.

File: pkg/varlinkapi/images.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (baude) [reply pkg/varlinkapi/images.go]: fixed

File: cmd/podman/commit.go (hunk 1 above)
Comment (rhatdan) [reply cmd/podman/commit.go]: Can we add these functions to pkg/util then?  I am all confused by this util?  Or change the name of the file to some other name.  Is this package uitl?

Comment (mheon) [reply cmd/podman/commit.go]: This is more of a podman-util (pkg/util/ is more libpod stuff). Though not that I think of it, we now have three packages named util in the project. We should do something about that.

Comment (rhatdan) [reply cmd/podman/commit.go]: If these are shared only by cmd in podman directory then it should go in the utils.go file.

Comment (mheon) [reply cmd/podman/commit.go]: They're in use by the Varlink package too, now.
Though we could still keep them in utils.go under Podman and just make them public? That sounds like a good compromise

Comment (baude) [reply cmd/podman/commit.go]: utils.go is in main ... i dont think that helps.

Comment (mheon) [reply cmd/podman/commit.go]: Oh. I forgot that Go forces a main package in addition to a main function...

Comment (rhatdan) [reply cmd/podman/commit.go]: Can we put util.go and utils.go into a subdir util and then put them in the same file package util?

Comment (baude) [reply cmd/podman/commit.go]: ok fixed up

**Comment by baude** [comment]
Fixed up per conversatio this morning.

**Comment by rhatdan** [comment]
@baude Already needs a rebase.

**Comment by mheon** [comment]
This LGTM once conflicts are resolved

**Comment by rhatdan** [comment]
LGTM
Fix conflicts, pass tests and we can merge.

**Comment by baude** [comment]
bot, retest this please

**Comment by baude** [comment]
@rh-atomic-bot r+

**Pull request closed without merging**